fix(distribution): enforce finite params in Gamma::new

oOOo-YKS opened the pull request:
- before:
    - Gamma::new raises an error oly when both shape and rate are infinite
- after:
    - Add infinite checks for shape/rate parameters
    - Add test cases for invalid infinite params

Closes #103
Refs: #98, #102

oOOo-YKS committed:
fix(distribution): enforce finite params in Gamma::new

- before:
    - Gamma::new raises an error oly when both shape and rate are zero
- after:
    - Add infinite checks for shape/rate parameters
    - Add test cases for invalid infinite params

Closes #103
Refs: #98, #102

oOOo-YKS commented:
sorry but I've made an error , it should be 
- before:
    -  Gamma::new raises an error only when both shape and rate are infinite

YeungOnion commented:
This is also the approach we took for Beta in #299, I think this is the least ambiguous route forward since our API is a floating point evaluation instead of a more thorough analysis approach.

Could you also,

1. deprecate the `GammaError::ShapeAndRateInfinite` variant?
1. update the `Display` impl for `GammaError` to mention finiteness restrictions?
1. drop the related checks for inf shape or rate in the other function implementations in this file?

oOOo-YKS committed:
fix(gamma): enforce finite params and cleanup validation

1. Deprecate GammaError::ShapeAndRateInfinite variant (no longer used)
2. Update error messages to explicitly mention NaN/infinite/zero/negative values:
  - "Shape must be finite (not NaN, infinite, zero, or negative)"
  - "Rate must be finite (not NaN, infinite, zero, or negative)"
3. Remove redundant infinity checks from cdf/sf methods (validation now centralized in constructor)
4. Related to #103, closes #98

oOOo-YKS commented:
For sure. I've finished all these three requirements and commit.

YeungOnion commented:
I've made one comment, but overall it looks good, thanks for the fix

YeungOnion committed:
Merge branch 'master' into fix/gamma-param-validation
